File: LinkedList/doubly/doubly.js
class LinkedListNode {
  constructor(value, next, previous) {
    this.value = value;
    this.next = next;
    this.previous = previous;
  }
}

/**
 * Implementation of a doubly linked list
 */
class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.length = 0;
  }

  pushFirst(value) {
    if (!this.head) return (this.head = new LinkedListNode(value));

    this.head = new LinkedListNode(value, this.head, null);
    this.head.next.previous = this.head;

    this.length++;

    return this;
  }

  getAt(index) {
    if (index < 0 || index > this.length) return null;

    let current = this.head;
    let count = 0;
    while (count < index) {
      current = current.next;
      count++;
    }
    return current.value;
  }

  log() {
    let output = "";
    let current = this.head;

    while (current) {
      output += `${current.value} <--> `;
      current = current.next;
    }
    console.log(`${output}null`);
  }
}

let list = new DoublyLinkedList();

list.pushFirst(5);
list.pushFirst(12);
list.pushFirst(23);
list.pushFirst(43);
list.pushFirst(87);

list.log();
// console.log(list.getAt(4));
// console.log(list.getAt(5));
// console.log(list.getAt(3));
// console.log(list.getAt(1));

File: LinkedList/doubly/doubly.go
package main

import "fmt"

//Implementation of the Doubly Linked List.
type LinkedListNode struct {
	value    int
	next     *LinkedListNode
	previous *LinkedListNode
}

type DoublyLinkedList struct {
	head   *LinkedListNode
	length int
}

func (l *DoublyLinkedList) pushFirst(value int) {
	previous := l.head
	l.head = &LinkedListNode{value: value}
	l.head.next = previous
	l.head.previous = nil
	l.length++
}

func (l DoublyLinkedList) getAt(index int) int {
	if index < 0 || index >= l.length {
		return -1
	}

	current := l.head
	for i := 0; i < index; i++ {
		current = current.next
	}

	return current.value
}

func (l DoublyLinkedList) log() {

	current := l.head
	for l.length != 0 {
		fmt.Printf("%d <--> ", current.value)
		current = current.next
		l.length--
	}

	fmt.Printf("null")
	fmt.Println()
}

func main() {
	list := DoublyLinkedList{}

	list.pushFirst(5)
	list.pushFirst(12)
	list.pushFirst(23)
	list.pushFirst(43)
	list.pushFirst(87)

	// list.log()
	fmt.Println(list.getAt(10))
}
